### PR TITLE
Improve deregister proposer code

### LINKED ISSUE #98: Prevent proposers from unregistering if they have pending blocks that could still be challenged
This is already flagged in the code as a TODO (see pic) 
image

One simple solution to this is to make proposers wait a week before they can 'deregister' (or at least, before they can withdraw their original bond). This is what optimism does (https://research.paradigm.xyz/optimism#incentives--bonds).

### PR DESCRIPTION
**Pull request by IlyasRidhuan**
This closes #98 as well as the comments in that issue.

This PR improves the handling of deregistering proposers:

1. In `deRegisterProposer`, we rotate proposer if the `currentProposer` is being deregistered and now correctly un-splice the deregistering proposer from the linked list.

2. A new `TimeLockedBond` struct is used to hold the bond of deregistering proposers that are waiting out the challenge period.

3. Bonds are now withdrawn using `withdrawBond` that checks that enough time has passed before a withdraw can be processed. A new api endpoint in `optimist` is used to trigger this.

### CODE REVIEW

**Comment by IlyasRidhuan** [comment]
This should be good to go after #109 is merged

**Review by Westlad — APPROVED** [review]

**Pull request merged**
